org.osgi.util.promise/src/org/osgi/util/promise/Promise.java

```diff
@@ -503,4 +503,166 @@
 	 * @since 1.2
 	 */
 	CompletionStage<T> toCompletionStage();
+
+	/**
+	 * Register a callback to be called with the failure for this Promise when
+	 * this Promise is resolved with a failure of a failure type. The callback
+	 * will not be called if this Promise is resolved successfully or if the
+	 * failure is not an instance of the specified failure type.
+	 * <p>
+	 * This method may be called at any time including before and after this
+	 * Promise has been resolved.
+	 * <p>
+	 * Resolving this Promise <i>happens-before</i> any registered callback is
+	 * called. That is, in a registered callback, {@link #isDone()} must return
+	 * {@code true} and {@link #getValue()} and {@link #getFailure()} must not
+	 * block.
+	 * <p>
+	 * A callback may be called on a different thread than the thread which
+	 * registered the callback. So the callback must be thread safe but can rely
+	 * upon that the registration of the callback <i>happens-before</i> the
+	 * registered callback is called.
+	 * 
+	 * @param failure The Consumer callback that receives the failure of this
+	 *            Promise if the failure is an instance of the specified failure
+	 *            type. Must not be {@code null}.
+	 * @param failureType The type of failure for which this callback will be
+	 *            called. If the failure is not an instance of the specified
+	 *            failure type, the specified callback will not be called. Must
+	 *            not be {@code null}.
+	 * @return This Promise.
+	 * @since 1.3
+	 */
+	Promise<T> onFailure(Consumer< ? super Throwable> failure,
```

Comment:
Introducing a dependency between the failureType and the failure consumer as follows, would eliminate the need to cast the generic throwable to the failure type within the consumer lambda in case the consumer requires to access any of the failureType's specific methods/fields.
```java
<F extends Throwable> Promise<T> onFailure(Consumer<? super F> failure, Class<F> failureType)
```

Reply:
`failureType` does not need to be a `Throwable`. It can be an interface which some set of custom exceptions implement. This is why it is `Class<?>` and not `Class<? extends Throwable>`. So we cannot establish a type linkage between the two parameters.

Reply:
I see why the class might not be a `Throwable`, but surely the consumer can be `? super X` to save the user casting every time?

Reply:
No, because F is unbound. There is no type-safe way to guarantee that the exception passed to the consumer can be cast to F's reification. 

`onFailure((RuntimeException e) -> e.printStackTrace(), Throwable.class);`

would be acceptable to the compiler, but at runtime any exception can be passed to the consumer which would fail when the (compiler generated) cast to RuntimeException fails.

Reply:
`<F> Promise<T> onFailure(Consumer<? super F> failure,	Class<? extends F> failureType);`

would be a possible signature choice since `F` is then bound by `failureType` and we can be confident that the exception can be passed to the consumer in a type safe manner.

But this only helps `onFailure` which directly consumes the exception. The other methods all get passed a Promise which supplies the exception as a Throwable and would need casting to any other exception type.

Reply:
> But this only helps onFailure which directly consumes the exception. The other methods all get passed a Promise which supplies the exception as a Throwable and would need casting to any other exception type.

Agreed, but I think it would be good to help here as we can.

Reply:
> `failureType` does not need to be a `Throwable`. It can be an interface which some set of custom exceptions implement. This is why it is `Class<?>` and not `Class<? extends Throwable>`. So we cannot establish a type linkage between the two parameters.

I did not consider that you might want to match on an interface implemented by the raised failure :). Perhaps this is an additional use-case to cover in the tests?

Reply:
> Perhaps this is an additional use-case to cover in the tests?

Good idea. Done!